### ISSUE TITLE
Mocking `spawn` (i.e. for tokio runtime)

### ISSUE THREAD

**Issue by MOZGIII** [asomers/mockall#96]
With this trait:
```rust
pub trait Runtime: Clone + Send + Sync + Unpin + Debug {
    fn spawn<F>(&self, future: F)
    where
        F: Future<Output = ()> + Send + 'static;
}
```

How do I mock `spawn`?

I tried like this:

```rust
let mock = MockRuntime::new();
mock.expect_spawn().returning(|_| ());
mock
```

```
   |
xx |             mock.expect_spawn().returning(|_| ());
   |                                            ^ consider giving this closure parameter a type
```

**Comment by asomers** [comment]
In this case `spawn` is a generic method, which really means it's an infinite collection of nongeneric methods.  You must set the expectation for the right version of spawn, or else the call won't match.  Assuming you're going to spawn a `BoxFuture`, you can do that like this.  Unfortunately, some `Future` types can be very long and hard to name.  There's no easy solution to the problem of figuring out a `Future`'s true type.
```rust
mock.expect_spawn::<BoxFuture<u32, ()>>().returning(|_| ());
```

Also, in this example you could use `return_const(())` instead of `returning(|_| ())`, but that won't fix the type problem.

**Comment by MOZGIII** [comment]
The problem is I'm using it with `async fn`, so it's flat out impossible to just name, afaik.
Good thing is I could implement the trait myself, and add a mocked method that's non-generic and simply receives the non-generic `BoxFuture`. This is far from perfect, but it works.

So, the issue is essence is really not about the `Future`s, but rather about taking compiler generated types, like `Fn`, `FnMut`, `FnOnce` and `Future` that can't be addressed any way other than `impl ...`.

**Comment by asomers** [comment]
I think this problem could be elegantly solved by #217 .
